File: android/app/src/main/java/com/cryptoproruto/CryptoProSignatureManager.kt
package com.cryptoproruto

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import java.io.File
import java.security.MessageDigest
import java.util.*

object CryptoProSignatureManager {
  fun signDocument(path: String, thumbprint: String?, pinCode: String?, detached: Boolean): WritableMap {
    val result = Arguments.createMap()

    try {
      val file = File(path)
      if (!file.exists()) {
        result.putBoolean("success", false)
        result.putString("error", "Document not found")
        result.putInt("errorCode", 404)
        return result
      }

      val data = file.readBytes()
      val signature = generateSignature(data, thumbprint)

      if (detached) {
        result.putBoolean("success", true)
        result.putString("signature", signature)
      } else {
        val signedPath = "$path.sig"
        File(signedPath).writeText(signature)
        result.putBoolean("success", true)
        result.putString("signedDocument", signedPath)
      }
    } catch (e: Exception) {
      result.putBoolean("success", false)
      result.putString("error", e.message)
      result.putInt("errorCode", -1)
    }

    return result
  }

  fun verifySignature(signedData: String): WritableMap {
    val result = Arguments.createMap()

    try {
      if (signedData.isEmpty()) {
        result.putBoolean("isValid", false)
        result.putString("error", "Invalid signature data")
        return result
      }

      result.putBoolean("isValid", true)

      val cert = Arguments.createMap()
      cert.putString("thumbprint", "00112233445566778899aabbccddeeff")
      cert.putString("subjectName", "CN=Test Signer")
      cert.putString("issuerName", "CN=Test CA")
      cert.putDouble("validFrom", System.currentTimeMillis().toDouble())
      cert.putDouble("validTo", (System.currentTimeMillis() + 31536000000).toDouble())
      cert.putString("serialNumber", "123456789")

      result.putMap("signerCertificate", cert)
      result.putDouble("signatureTime", System.currentTimeMillis().toDouble())
    } catch (e: Exception) {
      result.putBoolean("isValid", false)
      result.putString("error", e.message)
    }

    return result
  }

  private fun generateSignature(data: ByteArray, thumbprint: String?): String {
    val timestamp = System.currentTimeMillis() / 1000
    val digest = MessageDigest.getInstance("SHA-256")
    val hash = digest.digest(data).joinToString("") { "%02x".format(it) }
    return "$hash.$timestamp"
  }
}

File: ios/RNCryptoProRutoken/Managers/CryptoProSignatureManager.swift
import Foundation

class CryptoProSignatureManager {
  static let shared = CryptoProSignatureManager()

  private init() {}

  func signDocument(path: String, thumbprint: String?, pinCode: String?, detached: Bool) -> [String: Any] {
    do {
      guard FileManager.default.fileExists(atPath: path) else {
        return [
          "success": false,
          "error": "Document not found",
          "errorCode": 404
        ]
      }

      let documentData = try Data(contentsOf: URL(fileURLWithPath: path))
      let signature = generateSignature(from: documentData, thumbprint: thumbprint)

      if detached {
        return [
          "success": true,
          "signature": signature
        ]
      } else {
        let signedPath = path + ".sig"
        try signature.write(toFile: signedPath, atomically: true, encoding: .utf8)
        return [
          "success": true,
          "signedDocument": signedPath
        ]
      }
    } catch {
      return [
        "success": false,
        "error": error.localizedDescription,
        "errorCode": -1
      ]
    }
  }

  func verifySignature(_ signedData: String) -> [String: Any] {
    do {
      guard !signedData.isEmpty else {
        return [
          "isValid": false,
          "error": "Invalid signature data"
        ]
      }

      return [
        "isValid": true,
        "signerCertificate": [
          "thumbprint": "00112233445566778899aabbccddeeff",
          "subjectName": "CN=Test Signer",
          "issuerName": "CN=Test CA",
          "validFrom": Int(Date().timeIntervalSince1970) * 1000,
          "validTo": Int(Date().timeIntervalSince1970 + 31536000) * 1000,
          "serialNumber": "123456789"
        ],
        "signatureTime": Int(Date().timeIntervalSince1970) * 1000
      ]
    } catch {
      return [
        "isValid": false,
        "error": error.localizedDescription
      ]
    }
  }

  private func generateSignature(from data: Data, thumbprint: String?) -> String {
    let timestamp = String(Int(Date().timeIntervalSince1970))
    let hash = data.sha256Hash()
    return "\(hash).\(timestamp)"
  }
}

extension Data {
  func sha256Hash() -> String {
    var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    self.withUnsafeBytes {
      _ = CC_SHA256($0.baseAddress, CC_LONG(self.count), &digest)
    }
    return digest.map { String(format: "%02hhx", $0) }.joined()
  }
}

import CommonCrypto
